api/src/schemas.py

```diff
@@ -203,10 +203,76 @@ def is_accessible(cls, values):
 
 
 class CmrInput(WorkflowInputBase):
-    # cmr discovery
-    discovery: Literal[Discovery.cmr]
-
+    discovery: Literal["cmr"]
     version: Optional[str]
     include: Optional[str]
     temporal: Optional[List[datetime]]
     bounding_box: Optional[List[float]]
+
+
+# allows the construction of models with a list of discriminated unions
+ItemUnion = Annotated[Union[S3Input, CmrInput], Field(discriminator="discovery")]
+
+
+class Dataset(BaseModel):
+    collection: str
+    title: str
+    description: str
+    license: str
+    dashboard_is_periodic: bool
+    dashboard_time_density: str
+    spatial_extent: BboxExtent
+    temporal_extent: TemporalExtent
+    sample_files: List[str]  # unknown how this will work with CMR
+    discovery_items: List[ItemUnion]
+
+    class Config:
+        extra = Extra.allow
+
+    @root_validator
+    def check_time_density(cls, v):
+        if v["dashboard_is_periodic"] and v["dashboard_time_density"] not in [
+            "month",
+            "day",
+            "year",
+        ]:
+            raise ValueError("Invalid time density")
+        if not v["dashboard_is_periodic"] and v["dashboard_time_density"] != "null":
+            raise ValueError("Invalid time density")
+        return v
+
+    # collection id must be all lowercase, with optional - delimiter
+    @validator("collection")
+    def check_id(cls, v):
+        if not re.match(r"[a-z]+(?:-[a-z]+)*", v):
+            raise ValueError("Invalid id")
+        return v
+
+    # all sample files must begin with prefix and their last element must match regex
+    @root_validator
+    def check_sample_files(cls, v):
+        if "s3" not in [item.discovery for item in v["discovery_items"]]:
+            print("No s3 discovery items to validate sample files against")
+            return v
+        # TODO cmr handling/validation
+        valid_matches = []
+        for item in v["discovery_items"]:
+            if item.discovery == "s3":
+                valid_matches.append(
+                    {"prefix": item.prefix, "regex": item.filename_regex}
+                )
+        for fname in v["sample_files"]:
+            if not any([fname.startswith(match["prefix"]) for match in valid_matches]):
+                raise ValidationError(
+                    f"Invalid sample file - {fname} doesn't match prefix"
+                )
+            if not any(
+                [
+                    re.search(match["regex"], fname.split("/")[-1])
+                    for match in valid_matches
+                ]
+            ):
+                raise ValidationError(
+                    f"Invalid sample file - {fname} doesn't match provided regex"
+                )
+        return v
```

Comment:
this passes even if the prefix/filename_regex pairs are incorrect

this example passes even though it should fail because `foo/anotherrandomstring` will not be discovered by `valid_matches`

```
import re
valid_matches = []
for item in [{"prefix": "foo/", "filename_regex": "^randomstring*"}, {"prefix": "bar/", "filename_regex": "^anotherrandomstring*"}]:
    valid_matches.append(
        {"prefix": item["prefix"], "regex": item["filename_regex"]}
    )

for fname in ["foo/anotherrandomstring", "bar/randomstring"]:
    if not any([fname.startswith(match["prefix"]) for match in valid_matches]):
        print(fname, "error")
    if not any(
        [
            re.search(match["regex"], fname.split("/")[-1])
            for match in valid_matches
        ]
    ):
        print(fname, "error")